plugins/chat/app/services/chat/create_thread.rb

```diff
@@ -30,50 +30,47 @@ class CreateThread
     policy :threading_enabled_for_channel
     model :original_message
     transaction do
-      step :find_or_create_thread
+      model :thread, :find_or_create_thread
       step :associate_thread_to_message
-      step :fetch_membership
-      step :publish_new_thread
-      step :trigger_chat_thread_created_event
     end
+    model :membership, optional: true
+    step :publish_new_thread
+    step :trigger_chat_thread_created_event
 
     private
 
     def fetch_channel(params:)
       ::Chat::Channel.find_by(id: params.channel_id)
     end
 
-    def fetch_original_message(channel:, params:)
-      ::Chat::Message.find_by(id: params.original_message_id, chat_channel_id: params.channel_id)
-    end
-
     def can_view_channel(guardian:, channel:)
       guardian.can_preview_chat_channel?(channel)
     end
 
     def threading_enabled_for_channel(channel:)
-      channel.threading_enabled
+      channel.threading_enabled?
+    end
+
+    def fetch_original_message(channel:, params:)
+      ::Chat::Message.find_by(id: params.original_message_id, chat_channel_id: params.channel_id)
     end
 
     def find_or_create_thread(channel:, original_message:, params:)
-      if original_message.thread_id.present?
-        return context[:thread] = ::Chat::Thread.find_by(id: original_message.thread_id)
-      end
+      return original_message.thread if original_message.thread
 
-      context[:thread] = channel.threads.create(
+      channel.threads.create(
         title: params.title,
         original_message: original_message,
         original_message_user: original_message.user,
       )
-      fail!(context.thread.errors.full_messages.join(", ")) if context.thread.invalid?
```

Comment:
It's really nice that we can switch this from a step to a model and remove the need for this kind of error handling and manually setting the key in the context hash.